pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -630,22 +632,34 @@ func (c *Controller) getProxyServiceInstancesFromMetadata(proxy *model.Proxy) ([
 			if err != nil {
 				return nil, fmt.Errorf("failed to find target port for %v: %v", proxy.ID, err)
 			}
-			// Construct the ServiceInstance
-			out = append(out, &model.ServiceInstance{
-				Service:     modelService,
-				ServicePort: svcPort,
-				Endpoint: &model.IstioEndpoint{
-					Address:         proxy.IPAddresses[0],
-					EndpointPort:    uint32(targetPort),
-					ServicePortName: svcPort.Name,
-					// Kubernetes service will only have a single instance of labels, and we return early if there are no labels.
-					Labels:         proxy.WorkloadLabels[0],
-					ServiceAccount: svcAccount,
-					Network:        c.endpointNetwork(proxy.IPAddresses[0]),
-					Locality:       util.LocalityToString(proxy.Locality),
-					Attributes:     model.ServiceAttributes{Name: svc.Name, Namespace: svc.Namespace},
-				},
-			})
+			// Dedupe the target ports here - Service might have configured multiple ports to the same target port,
+			// we will have to create only one ingress listener/cluster per port so that we do not endup complaining about
+			// listener conflicts.
+			if _, exists := tps[targetPort]; !exists {
+				tps[targetPort] = svcPort
```

Comment:
Is this right? It seems it can not achieve what you described. When you have two different protocol named port(tcp-xx, http-xx).

Reply:
yes, this is not right. Please see the latest commit.